### ISSUE TITLE
Error when trying to calculate equilibrium with just a gaseous phase

### ISSUE THREAD

**Issue by martinvoigt** [reaktoro/reaktoro#65]
I wasn't able to calculate the equilibrium state of systems where just a gaseous phase would be present:

```Python
db = Database('supcrt98')
editor = ChemicalEditor(db)
editor.addGaseousPhase('H O')

system = ChemicalSystem(editor)

problem = EquilibriumProblem(system)
problem.setTemperature(150, 'celsius')
problem.setPressure(1,'bar')
problem.add('H2O', 1.0, 'kg')

state = equilibrate(problem)

print(state)
```

This is the error I got:
```
RuntimeError: 
****************************************************************************************
*** Error: Could not get a reference to a Phase instance with given index.
*** Reason: The given index 1 is out of bounds.
*** Location:  This error was encountered in :236.
****************************************************************************************
```

Seems there is no phase present at all... Playing around, sometimes I already got an error at `equilibrate(problem)`, telling me no equilibrium was found - but I couldn't reproduce this one again unfortunately.

I tried finding out more by trying to see what `state.properties().phaseAmounts()` returns, but wasn't able to get much out of it, as I didn't find much documentation on the `ChemicalVector` type yet.

It works if an aqueous phase is added using `editor.addAqueousPhase('H O')` to the problem though, so not sure if this is just a problem with displaying the output or with equilibrium calculations themselves.

Apart from this issue: Would it be possible to add an option to set the pressure to the vapor pressure at the specific temperature/composition when defining a problem? I guess this could be done by a user by running some iterations to find that phase boundary, but having it built in would be convenient I think...
Maybe that's already there, but I didn't find it in the docs yet.
I didn't think this through all the way, but this might require setting the volume to a fixed value (instead of the pressure) in such a problem. Is that possible, or will it ever be?

Generalizing this: How about adding some inverse-solving capabilities? An obvious interesting thing would be finding phase boundaries (such as water-vapor, as described above, but also between minerals etc.).

Another related "gimmick" would be to be able to define a certain pH for an equilibrium problem, which is realized by adding the right amount of an acid or base (such as HCl or NaOH). In PHREEQC, this can be done by defining a custom phase `H+` with an equilibrium constant of 0, and then forcing the saturation index of this phase to `-pH` using HCl, NaOH, or other chemicals. Being able to set saturation indices to a specific value (not only 0) is a useful capabilities in general as well...

**Comment by allanleal** [comment]
Hi @martinvoigt - thanks a lot for reporting this issue!

This is happening because of the current bug that the `ChemicalState` output method assumes there is always an aqueous phase. If you do instead:

~~~python
from reaktoro import *

db = Database('supcrt98')
editor = ChemicalEditor(db)
editor.addGaseousPhase('H O')

system = ChemicalSystem(editor)

problem = EquilibriumProblem(system)
problem.setTemperature(150, 'celsius')
problem.setPressure(1,'bar')
problem.add('H2O', 1.0, 'kg')

state = equilibrate(problem)

# print(state)
 
for s, n in zip(state.system().species(), state.speciesAmounts()):
    print( s.name(), n, "mol")
~~~

Then you'll see everything is fine, except the print statement. 

A more interesting calculation would be:

~~~python
from reaktoro import *

db = Database('supcrt98')
editor = ChemicalEditor(db)
editor.addGaseousPhase('H O C')

system = ChemicalSystem(editor)

problem = EquilibriumProblem(system)
problem.setTemperature(150, 'celsius')
problem.setPressure(1,'bar')
problem.add('O2', 1.0, 'mol')
problem.add('CH4', 1.0, 'mol')

state = equilibrate(problem)

for s, n in zip(state.system().species(), state.speciesAmounts()):
    print( s.name(), n, "mol")
~~~
which outputs:

~~~
C2H4(g) 5.060452970461436e-15 mol                                                                      
CH4(g) 0.4994021534786748 mol                                                                          
CO(g) 1.3109034927486926e-06 mol                                                                       
CO2(g) 0.5005965356178225 mol                                                                          
H2(g) 0.002390075181778016 mol                                                                         
H2O(g) 0.9988056178608625 mol                                                                          
O2(g) 1.5772989722659833e-22 mol 
~~~

So, here is the line of action: remove this presumption of existence of aqueous phase when printing a `ChemicalState` instance, and implement a more dedicated class for aqueous solution states that has a similar output as PHREEQC.

I'll reply the rest of your questions shortly. The `demos` dir has many examples on inverse equilibrium problems where pH and other properties can be fixed while titrating some unknown amounts of titrants (e.g., pH controlled by adding CO2, HCl, NaOH, etc.). There is also the possibility of having either one titrant or another. For example, fixed pH with either HCl or NaOH, and the calculation will figure out which one has to be added to avoid a problem without solution.

**Comment by allanleal** [comment]
> I tried finding out more by trying to see what state.properties().phaseAmounts() returns, but wasn't able to get much out of it, as I didn't find much documentation on the ChemicalVector type yet.

A `ChemicalVector` object has members `val`, `ddT`, `ddP`, and `ddn`. 

- `val` is the vector of values of the properties
- `ddT` is the vector of temperature derivatives of the properties
- `ddP` is the vector of pressure derivatives of the properties
- `ddn` is the matrix of composition (species amounts) derivatives of the properties

So, for example:

~~~python
ln_a = state.properties()lnActitivies()

print(ln_a.val)  # print the values
print(ln_a.ddn)  # print the derivatives with respect to speies amounts
~~~

**Comment by allanleal** [comment]
The requests below are valid and interesting:

> Would it be possible to add an option to set the pressure to the vapor pressure at the specific temperature/composition when defining a problem? 

> How about adding some inverse-solving capabilities? An obvious interesting thing would be finding phase boundaries (such as water-vapor, as described above, but also between minerals etc.)

I propose we continue the discussion of these two features in separate issues, for improved organization.

But just a few questions first. The vapor pressure: would it be the vapor pressure of pure water? As to phase boundaries, the existing inverse chemical equilibrium solver can impose any sort of constraint (there are existing ones, such as fixed pH, pe, etc., but one could also define custom constraints - more advanced usage though). It remains to code perhaps some additional constraints to make this phase boundary calculation more user friendly. What would be the constraints you need for phase boundaries?

**Comment by martinvoigt** [comment]
Many thanks for looking into the error, and pointing me to the demos - the `EquilibriumInverseProblem` is fantastic!

> But just a few questions first. The vapor pressure: would it be the vapor pressure of pure water? As to phase boundaries, the existing inverse chemical equilibrium solver can impose any sort of constraint (there are existing ones, such as fixed pH, pe, etc., but one could also define custom constraints - more advanced usage though). It remains to code perhaps some additional constraints to make this phase boundary calculation more user friendly. What would be the constraints you need for phase boundaries?

Vapor pressure of pure water is a good approximation in many cases, so this would already be great to have, but being able to fix pressure to the real boiling curve or phase boundaries in general would be great.

From what I can see, `EquilibriumInverseProblem` can be easily used to fix various compositional parameters (pH, eH, activities, phase amounts, ...) at constant P,T using titrants at the moment. Would it be possible to add constraints like equal chemical potentials of two or more phases at a given composition  (+- fixed P or T), and then inverse to find T and/or P where these constraints are met?

As a very simple example, to get the pressure on the vapor curve:

```Python
problem = EquilibriumInverseProblem(system)
problem.setTemperature(150, 'celsius')
problem.add('H2O', 1.0, 'kg')
problem.PhaseCoexistence('aqueous','gaseous')
```

I guess it might be necessary (or helpful) to be able to specify some initial pressure/temperature where the algorithm starts looking for a solution, as there can be multiple solutions (although the user might also want a list of all of them).

Of course it would be great if this would work for other phase types/combinations as well, e.g. coexistence of two minerals. The important constraint would be the equality of the chemical potentials of both phases, which defines the phase boundary.

An interesting case is also the NaCl-H2O system, which has lots of application in hydrothermal systems. It would be great to be able to calculate high T/P thermodynamics of the system, but I guess at the moment this limited by not having NaCl as a vapor species in the databases. Out of curiosity: Would it be possible to define additional species in reaktoro based on custom/empirical equation of state/thermodynamic properties? I'm asking because there's a [nice empirical model](https://www.sciencedirect.com/science/article/pii/S0016703707002943) of the NaCl-H2O system, but the equations describing the thermodynamic properties are not in a standard form, but empirical equations for various properties. How flexible is reaktoro with adding such data?

Another example of what would be great for the inverse solver would be fixing the total volume of a system (at least I didn't find this in the methods yet). As an example:

```Python
problem = EquilibriumInverseProblem(system)
problem.setTemperature(150, 'celsius')
problem.add('H2O', 1.0, 'kg')
problem.setV(1,'L')
```

Which would then calculate the pressure in the system and the equilibrium state. Not sure how easy this would be to implement, I guess done properly this would require minimizing the [Helmholz Free Energy](https://en.wikipedia.org/wiki/Helmholtz_free_energy) instead of Gibbs. But maybe there's an easy way to formulate this as an inverse problem without opening that Pandora's box?

**Comment by allanleal** [comment]
Thanks @martinvoigt !

This is how the equilibrium constraint function for fixing the activity of a species with index `i` is implemented (this is then used as a base to implement a constraint function for pH, by choosing species `H+`, knowing that `pH = -log10(activity[H+])`):

~~~c++
EquilibriumConstraint f = [=](VectorConstRef x, const ChemicalState& state) mutable
{
    ln_ai = state.properties().lnActivities()[ispecies];
    return ln_ai - ln_val;
};
~~~

Above `ln_val` is a captured variable that is the value specified by the user. All equilibrium constraint functions are defined as a residual (above, we had `current property value - desired property value`). For equal chemical potentials of two species (e.g., H2O(aq) and H2O(g)):

~~~c++
EquilibriumConstraint f = [=](VectorConstRef x, const ChemicalState& state) mutable
{
    u1 = state.properties().chemicalPotentials()[ispecies1];
    u2 = state.properties().chemicalPotentials()[ispecies2];
    return u1 - u2;
};
~~~

I'll reply the rest shortly.

**Comment by allanleal** [comment]
@martinvoigt this issue has been fixed at #66 

There is now a new release 1.0.2 in conda (possibly in a hour or so because of some delay in conda-forge) with this bug fix.

**Comment by allanleal** [comment]
Let's continue the other conversation in another dedicated issue.
